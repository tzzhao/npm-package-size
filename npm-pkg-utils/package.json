{
  "name": "npm-pkg-utils",
  "version": "0.0.1",
  "description": "Bundle and minify npm package, and compute its gzipped size",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc",
    "test": "echo \"Error: no test specified\" && exit 1",
    "watch:dev": "nodemon",
    "watch": "tsc-watch --onSuccess \"node ./dist/index.js",
    "start": "yarn build && node dist/index.js"
  },
  "author": "tzhao",
  "license": "ISC",
  "devDependencies": {
    "@types/node": "^14.10.1",
    "@types/npm": "^2.0.31",
    "@types/semver": "^7.3.3",
    "eslint": "^7.7.0",
    "nodemon": "^2.0.4",
    "tsc-watch": "^4.2.9",
    "ts-node": "^9.0.0",
    "typescript": "^4.0.2"
  },
  "dependencies": {
    "autoprefixer": "^9.7.6",
    "builtin-modules": "^3.1.0",
    "css-loader": "^4.2.1",
    "csso-webpack-plugin": "^2.0.0-beta.1",
    "escape-string-regexp": "^4.0.0",
    "file-loader": "^6.0.0",
    "memory-fs": "^0.5.0",
    "mini-css-extract-plugin": "^0.10.0",
    "npm": "^6.14.8",
    "sanitize-filename": "^1.6.3",
    "sass-loader": "^9.0.3",
    "semver": "^7.3.2",
    "shebang-loader": "^0.0.1",
    "terser-webpack-plugin": "^2.3.5",
    "webpack": "^4.44.1"
  }
}
