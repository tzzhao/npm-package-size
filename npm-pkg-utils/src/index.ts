import {buildPackage, getVersions} from './build-package';

export * from './build-package';
export * from './command-utils';
export * from './interfaces';

export async function getLatestVersionsAndSize(packageName: string) {
  return getVersions(packageName).then(async (versions) => {
    const buildPromises = versions.map(async (version) => {
      try {
        const result = await buildPackage(packageName, version);
        console.log(result);
        return result;
      } catch (e) {
        console.error(e);
        return e;
      }
    });
    return await Promise.all(buildPromises);
  });
}

// const packageName: string = '@angular/core';
//
// getLatestVersionsAndSize(packageName).then(
//     (results) => {
//       console.log('Processed finished');
//     }
// );
