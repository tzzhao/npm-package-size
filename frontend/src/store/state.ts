import {PackageInformation} from 'npm-pkg-utils';

export enum PackageState {
  INIT, LOADING, READY, ERROR
}

export interface RootState {
  state: PackageState,
  packageInfos: PackageInformation[],
  globalError: Error | undefined
}

export const initialState: RootState = {
  state: PackageState.INIT,
  packageInfos: [],
  globalError: undefined
};
