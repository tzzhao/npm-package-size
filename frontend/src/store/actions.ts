import {PackageInformation} from 'npm-pkg-utils';
import {PackageState} from './state';

export const SET_LOADING = 'SET_LOADING';
export const SET_PACKAGE_INFOS = 'SET_PACKAGE_INFOS';
export const SET_ERROR = 'SET_ERROR';

export const LoadingAction = () => ({
  type: SET_LOADING,
  payload: {
    state: PackageState.LOADING,
    globalError: undefined,
    packageInfos: []
  }
});

export const SetPackageInformationAction = (packageInfos: PackageInformation[]) => ({
  type: SET_PACKAGE_INFOS,
  payload: {
    state: PackageState.READY,
    globalError: undefined,
    packageInfos
  }
});

export const SetErrorAction = (globalError: Error) => ({
  type: SET_ERROR,
  payload: {
    state: PackageState.ERROR,
    globalError,
    packageInfos: []
  }
});
