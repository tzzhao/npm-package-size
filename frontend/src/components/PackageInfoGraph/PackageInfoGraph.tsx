import {PackageError, PackageInformation} from 'npm-pkg-utils';
import * as React from 'react';
import {connect} from 'react-redux';
import {RootState} from '../../store/state';
import {getDisplaySize} from '../../utils/size';
import {BarProperties} from '../BarGraph/Bar';
import {BarGraph} from '../BarGraph/BarGraph';

export interface PackageInfoGraphProperties {
  packagesInformation: PackageInformation[]
}

const PackageInfoGraphNotConnected: React.FC<Partial<PackageInfoGraphProperties>> = props => {
    const pkgInfos = props.packagesInformation!.filter(info => info.size);

    const {biggestPackageSize, failedPackages} = getBiggestPackageSizeAndFailedPackages(props.packagesInformation!);

    const failedPackagesMessage =
        failedPackages.length > 0 ? `An issue occurred while processing the following packages: ${failedPackages.join(', ')}` : '';

    const barsProps = computeBarProperties(pkgInfos, biggestPackageSize);

    return (
        <div>
          <BarGraph barsProps={barsProps} title={props.packagesInformation![0]?.pkgName}/>
          {failedPackagesMessage ? <div className="bar-graph-error">{failedPackagesMessage}</div> : ''}
        </div>
    );
};

function getBiggestPackageSizeAndFailedPackages(packagesInformation: PackageInformation[])
    : {biggestPackageSize: number, failedPackages: string[]} {
  const failedPackages: string[] = [];
  let maxSize = 0;
  for (const pkgInfo of packagesInformation) {
    const pkgSize: number = pkgInfo.size;
    if (!pkgSize) {
      failedPackages.push(`${pkgInfo.pkgName}@${pkgInfo.pkgVersion} (${(pkgInfo as any as PackageError).name})`);
    } else {
      maxSize = Math.max(pkgSize, maxSize);
    }
  }
  return {biggestPackageSize: maxSize, failedPackages};
}

function computeBarProperties(pkgInfos: PackageInformation[], maxSize: number): BarProperties[] {
    return pkgInfos
      .reduce((barsProps: BarProperties[], currentBarProp: PackageInformation) => {
        const name = currentBarProp.pkgVersion;
        const size: number = currentBarProp.size;
        barsProps.push(
            {
              height: `${size * 100 / maxSize}%`,
              description: `Minified size: ${getDisplaySize(size)}`,
              name: `${name} (minified)`,
              color: '#2B547E'
            }
        );

        const gzipSize: number = currentBarProp.gzip;
        barsProps.push(
            {
              height: `${gzipSize * 100 / maxSize}%`,
              description: `Gzipped size: ${getDisplaySize(gzipSize)}`,
              name: `${name} (gzipped)`,
              color: '#659EC7'
            }
        );
        return barsProps;
      }, []);
}

const mapStateToPackageInfoGraphProps  = (state: RootState) => {
  return {
    packagesInformation: state.packageInfos
  };
};

export const PackageInfoGraph = connect(mapStateToPackageInfoGraphProps)(PackageInfoGraphNotConnected);
