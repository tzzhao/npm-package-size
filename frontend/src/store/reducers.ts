import {Reducer} from 'redux';
import {Action, SET_ERROR, SET_LOADING, SET_PACKAGE_INFOS} from './actions';
import {initialState, RootState} from './state';

export const reducer : Reducer<RootState, Action> = ((state: RootState = initialState, action: Action) => {
  switch(action.type) {
    case SET_LOADING:
      return {...state, ...action.payload};
    case SET_PACKAGE_INFOS:
      return {...state, ...action.payload};
    case SET_ERROR:
      return {...state, ...action.payload};
    default:
      return state;
  }
});
