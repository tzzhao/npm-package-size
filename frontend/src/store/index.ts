import {createStore, Store} from 'redux';
import {reducer} from './reducers';
import {initialState} from './state';

export const store: Store = createStore(reducer, initialState);
