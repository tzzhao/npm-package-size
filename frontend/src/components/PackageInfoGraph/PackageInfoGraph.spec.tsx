import {render, RenderResult} from '@testing-library/react';
import {PackageError, PackageInformation} from 'npm-pkg-utils';
import * as React from 'react';
import {act} from 'react-test-renderer';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import {createStore, Store} from 'redux';
import {SetPackageInformationAction} from '../../store/actions';
import {reducer} from '../../store/reducers';
import {initialState} from '../../store/state';
import {
  TEST_PACKAGE_INFO_1,
  TestSetEmptyPackageInformationAction,
  TestSetPackageInformationAction,
} from '../../utils/test/store/actions';
import {PackageInfoGraph} from './PackageInfoGraph';

const store: Store = createStore(reducer, initialState);

test('Test package info graph is empty when package infos are empty', () => {
  act(() => {
    store.dispatch(TestSetEmptyPackageInformationAction);
  });

  const renderResult: RenderResult = render(<Provider store={store}><PackageInfoGraph /></Provider>);
  expect(renderResult.baseElement.textContent).toBe("");
});

test('Test package info graph with 2 package information - SNAPSHOT', () => {
  act(() => {
    store.dispatch(TestSetPackageInformationAction);
  });

  const component = renderer.create(<Provider store={store}><PackageInfoGraph /></Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test package info graph with 1 missing package information - SNAPSHOT', () => {
  const MISSING_PACKAGE_ERROR: PackageError = {
    pkgName: 'package',
    pkgVersion: '1.0.2',
    name: 'NpmError',
    message: 'Some error message'
  };

  act(() => {
    store.dispatch(SetPackageInformationAction([TEST_PACKAGE_INFO_1, MISSING_PACKAGE_ERROR as any as PackageInformation]));
  });

  const component = renderer.create(<Provider store={store}><PackageInfoGraph /></Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
