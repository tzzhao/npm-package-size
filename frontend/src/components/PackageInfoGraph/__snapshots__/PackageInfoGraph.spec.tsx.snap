// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test package info graph with 1 missing package information 1`] = `
<div>
  <div
    className="bar-graph"
  >
    <div
      className="bars-container"
    >
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#2B547E",
            "height": "100%",
          }
        }
        title="Minified size: 15.20kB"
      />
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#659EC7",
            "height": "0.7893178977833323%",
          }
        }
        title="Gzipped size: 120B"
      />
    </div>
    <div
      className="bars-legend-container"
    >
      <div
        className="bar-legend"
      >
        package@1.0.1 (minified)
      </div>
      <div
        className="bar-legend"
      >
        package@1.0.1 (gzipped)
      </div>
    </div>
  </div>
  <div
    className="bar-graph-error"
  >
    An issue occurred while processing the following packages: package@1.0.2
  </div>
</div>
`;

exports[`Test package info graph with 2 package information 1`] = `
<div>
  <div
    className="bar-graph"
  >
    <div
      className="bars-container"
    >
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#2B547E",
            "height": "0.009986783064441788%",
          }
        }
        title="Minified size: 15.20kB"
      />
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#659EC7",
            "height": "0.00007882746614043377%",
          }
        }
        title="Gzipped size: 120B"
      />
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#2B547E",
            "height": "100%",
          }
        }
        title="Minified size: 152.23MB"
      />
      <div
        aria-describedby={null}
        className="bar"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#659EC7",
            "height": "0.08034489486363712%",
          }
        }
        title="Gzipped size: 122.31kB"
      />
    </div>
    <div
      className="bars-legend-container"
    >
      <div
        className="bar-legend"
      >
        package@1.0.1 (minified)
      </div>
      <div
        className="bar-legend"
      >
        package@1.0.1 (gzipped)
      </div>
      <div
        className="bar-legend"
      >
        package@1.2.1 (minified)
      </div>
      <div
        className="bar-legend"
      >
        package@1.2.1 (gzipped)
      </div>
    </div>
  </div>
  
</div>
`;
